core/state_transition.go

```diff
@@ -373,19 +367,13 @@ func (st *StateTransition) StakingTransitionDb() (usedGas uint64, err error) {
 	return st.gasUsed(), err
 }
 
-func (st *StateTransition) applyCreateValidatorTx(createValidator *staking.CreateValidator, blockNum *big.Int) error {
-	if createValidator.Amount.Sign() == -1 {
-		return errNegativeAmount
-	}
-
-	if val := createValidator.ValidatorAddress; st.state.IsValidator(val) {
-		return errors.Wrapf(errValidatorExist, common2.MustAddressToBech32(val))
-	}
-
-	if !CanTransfer(st.state, createValidator.ValidatorAddress, createValidator.Amount) {
-		return errInsufficientBalanceForStake
+func (st *StateTransition) verifyAndApplyCreateValidatorTx(
+	createValidator *staking.CreateValidator, blockNum *big.Int,
+) error {
+	err := VerifyCreateValidatorStakingMsg(st.state, st.evm.EpochNumber, createValidator)
```

Comment:
You can return the created validator from VerifyCreateValidatorStakingMsg (maybe rename to VerifyAndCreateValidatorStakingMsg), and directly commit with that validator. Then you don't need to redo the whole wrapper.validator = .... ; wrapper.snapshot = .... thing. They are duplicate code.

Reply:
alright, refactored the checks to return components that need to be updated in the state